staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -903,6 +968,12 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("uniqueItems"), "uniqueItems cannot be set to true since the runtime complexity becomes quadratic"))
 	}
 
+	if len(schema.XValidations) > 0 {
+		// This schema will need Schema information to validate the validation rules, so we construct now
+		// so that it can also be used by the validation rules of any child schemas.
+		nodeCostInfo.LoadSchemaIfAbsent(schema, isRoot)
```

Comment:
are we relying on calls in the right order to populate this data to let us avoid doing work twice? do we have a way to verify how many times we're doing the ~expensive schema calculations so we can make sure any ordering changes in calls don't reintroduce expensive paths?

Reply:
I've eliminated this call.  The flow should be easier to reason about now.